envoy/ssl/handshaker.h

```diff
@@ -156,5 +165,104 @@ class HandshakerFactory : public Config::TypedFactory {
   virtual SslCtxCb sslctxCb(HandshakerFactoryContext& handshaker_factory_context) const PURE;
 };
 
+struct SelectionResult {
+  enum class SelectionStatus {
+    // Continue the TLS handshake.
+    Continue,
+    // Block the TLS handshake.
+    Stop,
+    // Abort the TLS handshake.
+    AbortHandshake,
+  };
+  // If the value is Pending, the selection is asynchronous.
+  SelectionStatus status;
+  // It will be nullptr when status is Stop or AbortHandshake.
+  const Ssl::TlsContext* selected_ctx;
+  // Enable OCSP stapling response when it is true.
+  bool staple;
+};
+
+/**
+ * Used to return the result from an synchronous/asynchronous cert selection.
+ */
+class CertSelectionCallback {
+public:
+  virtual ~CertSelectionCallback() = default;
+
+  virtual Event::Dispatcher& dispatcher() PURE;
+
+  /**
+   * Called when the asynchronous cert selection completes.
+   * @param selected_ctx selected Ssl::TlsContext, it's empty when selection failed.
+   * @param staple true when need to set OCSP response.
+   */
+  virtual void onCertSelectionResult(OptRef<const Ssl::TlsContext> selected_ctx, bool staple) PURE;
+};
+
+using CertSelectionCallbackPtr = std::unique_ptr<CertSelectionCallback>;
+
+class TlsCertificateSelector {
+public:
+  virtual ~TlsCertificateSelector() = default;
+
+  /**
+   * select TLS context based on the client hello.
+   */
+  virtual SelectionResult selectTlsContext(const SSL_CLIENT_HELLO* ssl_client_hello,
+                                           CertSelectionCallbackPtr cb) PURE;
+};
+
+using TlsCertificateSelectorPtr = std::unique_ptr<TlsCertificateSelector>;
+
+class ContextSelectionCallback {
+public:
+  virtual ~ContextSelectionCallback() = default;
+
+  /**
+   * @return reference to the initialized Tls Contexts.
+   */
+  virtual const std::vector<TlsContext>& getTlsContexts() const PURE;
+};
+
+using ContextSelectionCallbackWeakPtr = std::weak_ptr<ContextSelectionCallback>;
```

Comment:
Please get rid of the weak_ptr. I think you can ensure the lifetime of the affected objects are correct. The `ServerContextImpl` should own any objects used for cert selection, so it can pass a reference to itself that will be valid for the required lifetime.